File: slide-engine/src/key-pool.ts
/**
 * Pool xoay vòng nhiều API key pinkyne cho slide-engine — khi model config
 * đang active gặp lỗi quota/401/403/429, tự chuyển sang key kế tiếp còn dùng
 * được thay vì để mọi lần sinh sau đó đều thất bại.
 *
 * Đọc từ OPENAI_API_KEYS (phân tách dấu phẩy) trong backend/.env; nếu không
 * có thì chỉ có 1 key (tương thích ngược — không bắt buộc phải cấu hình pool).
 */

const COOLDOWN_MS = 5 * 60 * 1000 // 5 phút

class KeyPool {
  private keys: string[]
  private badUntil = new Map<string, number>()
  private idx = 0

  constructor(keys: string[]) {
    this.keys = keys
  }

  size(): number {
    return this.keys.length
  }

  current(): string | null {
    if (this.keys.length === 0) return null
    const now = Date.now()
    for (let i = 0; i < this.keys.length; i++) {
      const k = this.keys[this.idx % this.keys.length]
      if ((this.badUntil.get(k) || 0) <= now) return k
      this.idx++
    }
    // Tất cả đang cooldown — vẫn trả về key ít mới hỏng nhất để thử (còn hơn không key nào)
    return this.keys.reduce((best, k) => ((this.badUntil.get(k) || 0) < (this.badUntil.get(best) || 0) ? k : best))
  }

  /** Đánh dấu key vừa lỗi (cooldown 5 phút), xoay sang key kế tiếp. Trả về key kế tiếp (nếu có). */
  markBad(key: string, reason = ''): string | null {
    this.badUntil.set(key, Date.now() + COOLDOWN_MS)
    this.idx++
    const now = Date.now()
    const ready = this.keys.filter((k) => (this.badUntil.get(k) || 0) <= now).length
    console.log(`[key-pool] key ...${key.slice(-6)} tạm khóa 5 phút (${reason}) — còn ${ready}/${this.keys.length} key sẵn sàng`)
    return this.current()
  }
}

let pool: KeyPool | null = null

export function getKeyPool(): KeyPool {
  if (!pool) {
    const raw = process.env.OPENAI_API_KEYS || ''
    let keys = raw
      .split(',')
      .map((k) => k.trim())
      .filter(Boolean)
    if (keys.length === 0) {
      const single = process.env.OPENAI_API_KEY || ''
      keys = single ? [single] : []
    }
    pool = new KeyPool(keys)
    if (keys.length > 1) console.log(`[key-pool] đã nạp ${keys.length} key pinkyne để tự xoay khi hết quota`)
  }
  return pool
}

/** Heuristic: message lỗi này có phải do hết quota/key hỏng không (nên xoay key)? */
export function isQuotaError(message: string): boolean {
  const m = (message || '').toLowerCase()
  return ['quota', 'insufficient', 'billing', 'exhausted', 'rate limit', 'token remain'].some((k) => m.includes(k))
}

File: backend/key_pool.py
"""
Pool xoay vòng nhiều API key pinkyne — khi 1 key hết quota/bị lỗi 401/403/429,
tự chuyển sang key kế tiếp còn dùng được thay vì làm cả luồng thất bại.

Đọc từ OPENAI_API_KEYS (phân tách dấu phẩy) trong .env; nếu không có thì dùng
OPENAI_API_KEY đơn lẻ (tương thích ngược, không cần đổi .env để chạy được).
"""
import os
import time
import threading

_COOLDOWN_SECONDS = 300  # 5 phút — đủ để quota có thể hồi phục, tránh đập liên tục vào key vừa hỏng


class KeyPool:
    def __init__(self, keys: list[str]):
        self._keys = keys
        self._bad_until: dict[str, float] = {}
        self._idx = 0
        self._lock = threading.Lock()

    def current(self) -> str | None:
        """Trả về key hiện dùng được (bỏ qua các key đang cooldown)."""
        with self._lock:
            if not self._keys:
                return None
            n = len(self._keys)
            now = time.time()
            for _ in range(n):
                k = self._keys[self._idx % n]
                if self._bad_until.get(k, 0) <= now:
                    return k
                self._idx += 1
            # Tất cả đều đang cooldown — vẫn trả về key ít mới hỏng nhất để thử (còn hơn không key nào)
            return min(self._keys, key=lambda k: self._bad_until.get(k, 0))

    def mark_bad(self, key: str, reason: str = "") -> str | None:
        """Đánh dấu key vừa lỗi (cooldown 5 phút), xoay sang key kế tiếp. Trả về key kế tiếp (nếu có)."""
        with self._lock:
            self._bad_until[key] = time.time() + _COOLDOWN_SECONDS
            self._idx += 1
            now = time.time()
            ready = sum(1 for k in self._keys if self._bad_until.get(k, 0) <= now)
            tail = key[-6:] if len(key) >= 6 else key
            print(f"[key-pool] key ...{tail} tạm khóa 5 phút ({reason}) — còn {ready}/{len(self._keys)} key sẵn sàng")
        return self.current()

    def size(self) -> int:
        return len(self._keys)


_pool: KeyPool | None = None


def get_pool() -> KeyPool:
    global _pool
    if _pool is None:
        raw = os.getenv("OPENAI_API_KEYS", "").strip()
        keys = [k.strip() for k in raw.split(",") if k.strip()]
        if not keys:
            single = os.getenv("OPENAI_API_KEY", "").strip()
            keys = [single] if single else []
        _pool = KeyPool(keys)
        if len(keys) > 1:
            print(f"[key-pool] đã nạp {len(keys)} key pinkyne để tự xoay khi hết quota")
    return _pool


def is_quota_error(status_code: int | None, message: str) -> bool:
    """Heuristic dùng chung: lỗi này có nên xoay sang key/provider khác không?"""
    if status_code in (401, 402, 403, 429):
        return True
    msg = (message or "").lower()
    return any(k in msg for k in ("quota", "insufficient", "billing", "exhausted", "rate limit", "token remain"))
